### PR TITLE
[WPF][XamMac][GtkMac] Enable formatted text in Buttons

### PR DESCRIPTION
**Pull request by mcumming**
Adding MenuButton support for formatted text to be used in MenuItems and the selected item. 

Backends:
- [x] WPF
- [x] GtkMac
- [x] XamMac

### CODE REVIEW

File: Xwt.WPF/Xwt.WPFBackend/ButtonBackend.cs [hunk 1]
```diff
@@ -100,39 +101,50 @@ protected ButtonBackend (ButtonBase impl)
...
-					Button.Content = accessText;
-				else
-					Button.Content = accessText.Text.Replace ("_", "__");
-			else {
+			accessText.Text = label ?? string.Empty;
+			var mnemonicValue = useMnemonic ? accessText.Text : accessText.Text.Replace ("_", "__");
+
+			if (image.IsNull) {
+				var formattedText = FormattedText.FromMarkup (mnemonicValue);
+				if (mnemonicValue.CompareTo (formattedText.Text) != 0) {
```
Comment (Therzok) [inline Xwt.WPF/Xwt.WPFBackend/ButtonBackend.cs]:
```suggestion
				if (!mnemonicValue.Equals (formattedText.Text)) {
```

File: Xwt.WPF/Xwt.WPFBackend/ButtonBackend.cs [hunk 2]
```diff
@@ -62,35 +62,36 @@ protected ButtonBackend (ButtonBase impl)
 			get { return (IButtonEventSink)base.EventSink; }
 		}
 
-		public void SetButtonStyle (ButtonStyle style) {
```
Comment (sevoku) [inline Xwt.WPF/Xwt.WPFBackend/ButtonBackend.cs]: Please avoid unrelated formatting changes

File: Xwt.WPF/Xwt.WPFBackend/MenuItemBackend.cs [hunk 3]
```diff
@@ -35,11 +35,12 @@
...
 using SWMI = System.Windows.Media.Imaging;
 using Xwt.Backends;
-
+using Xwt.WPFBackend.Utilities;
+using WindowsOrientation = System.Windows.Controls.Orientation;
 
 namespace Xwt.WPFBackend
 {
-	public class MenuItemBackend : Backend, IMenuItemBackend
+	public class MenuItemBackend : WidgetBackend, IMenuItemBackend
```
Comment (sevoku) [inline Xwt.WPF/Xwt.WPFBackend/MenuItemBackend.cs]: Why is this needed?

File: Xwt.Gtk/Xwt.GtkBackend/MenuItemBackend.cs [hunk 4]
```diff
@@ -50,6 +50,11 @@ public MenuItemBackend (Gtk.MenuItem item)
 		{
 			this.item = item;
 			label = (Gtk.Label) item.Child;
+			if (label != null) {
+				label.Realized += HandleStyleUpdate;
+				label.StyleSet += HandleStyleUpdate;
```
Comment (netonjm) [inline Xwt.Gtk/Xwt.GtkBackend/MenuItemBackend.cs]: this should be unsubscribed on disposing ?

File: Xwt/Xwt.Backends/IButtonBackend.cs [hunk 5]
```diff
@@ -36,6 +36,12 @@ public interface IButtonBackend: IWidgetBackend
 		void SetButtonStyle (ButtonStyle style);
 		void SetButtonType (ButtonType type);
 		void SetContent (string label, bool useMnemonic, ImageDescription image, ContentPosition position);
+
+		/// <summary>
+		/// Sets the formatted text display for this menu item.
+		/// </summary>
+		/// <param name="text">The text to display with additional markup for formatting <see cref="Xwt.FormattedText"/></param>
+		void SetFormattedText (FormattedText text);
```
[the remark this answers is not in the document]
Comment (mcumming) [reply Xwt/Xwt.Backends/IButtonBackend.cs]: I don't think it would make much of a difference on the implementation of the backend either way. I think what would make it easier, was if all the labels were converted to FormattedText objects then applied. But you would need a way to determine of the FormattedText did not in fact contain any formatting so that you could short curcuit the label creation and styling for plain text.

[the remark this answers is not in the document]
Comment (sevoku) [reply Xwt/Xwt.Backends/IButtonBackend.cs]: well, it'd let us avoid double-init here: https://github.com/mono/xwt/pull/959/files#diff-75a1dc4c07ad775d27eb0deadeee862fR255. Whether FormattedText contains formatting or not can be determined by `FormattedText.Attributes.Count`.

[the remark this answers is not in the document]
Comment (mcumming) [reply Xwt/Xwt.Backends/IButtonBackend.cs]: So my reasoning for setting it up that way was that if you change the `Label` of the button, it calls `SetContent`, which recreates the `labelWidget` internally. So setting the `Markup` property it should do all the things that setting the `Label` does, plus applying the proper formatting. This is why there is a check to see if the `labelWidget` already exists in `SetFormattedContent` instead of just recreating it every time. In this way, `SetFormattedText` uses `SetContent` to configure everything properly, and then `SetFormattedContent` extends that to add the proper formatting.

File: Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs [hunk 6]
```diff
@@ -220,7 +249,13 @@ public void SetButtonType (ButtonType type)
 			Button b = (Button) Frontend;
 			SetContent (b.Label, b.UseMnemonic, image, b.ImagePosition);
 		}
-		
+
+		public void SetFormattedText (FormattedText text)
+		{
+			SetContent (text.Text, Widget.UseUnderline, this.image, ContentPosition.Center);
```
Comment (netonjm) [inline Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs]: possible nre in text

File: Xwt.Gtk/Xwt.GtkBackend/MenuItemBackend.cs [hunk 7]
```diff
@@ -168,7 +168,25 @@ void ImageMenuItemStateChanged (object o, Gtk.StateChangedArgs args)
 				}
 			}
 		}
-		
+
+		FormattedText formattedText = null;
+		public void SetFormattedText (FormattedText text)
+		{
+			label.Text = text.Text;
```
Comment (netonjm) [inline Xwt.Gtk/Xwt.GtkBackend/MenuItemBackend.cs]: possible nre in text

File: Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs [hunk 8]
```diff
@@ -272,6 +307,15 @@ void HandleSizeAllocated (object o, Gtk.SizeAllocatedArgs args)
 			Widget.Child.SizeAllocate (args.Allocation);
 			args.RetVal = true;
 		}
+
+		protected override void Dispose (bool disposing)
+		{
+			if (labelWidget != null) {
+				labelWidget.Realized -= HandleStyleUpdate;
```
Comment (sevoku) [inline Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs]: `labelWidget` is created either in `SetFormattedText()` (events subscribed) or in `SetContent ()` (events not subscribed), so this will thrwo NRE for regular text (SetFormattedText() not called).

File: Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs [hunk 9]
```diff
@@ -272,6 +307,15 @@ void HandleSizeAllocated (object o, Gtk.SizeAllocatedArgs args)
 			Widget.Child.SizeAllocate (args.Allocation);
 			args.RetVal = true;
 		}
+
+		protected override void Dispose (bool disposing)
+		{
+			if (labelWidget != null) {
```
Comment (sevoku) [inline Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs]:
```suggestion
			if (disposing && labelWidget != null) {
```
and set `labelWidget = null` at the end of the block to not NRE on double disposing.

File: Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs [hunk 10]
```diff
@@ -197,6 +197,35 @@ public void SetContent (string label, bool useMnemonic, ImageDescription image,
...
+			if (labelWidget == null) {
+				labelWidget = new Gtk.Label (label.Text);
+				Widget.Label = null;
+				Widget.Image = labelWidget;
+			}
+
+			formattedText = label;
+			labelWidget.ApplyFormattedText (formattedText);
+			labelWidget.ShowAll ();
+			labelWidget.Realized += HandleStyleUpdate;
```
Comment (sevoku) [inline Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs]: Will subscribe on every call, what if the Markup property get's updated?

File: Xwt.Gtk/Xwt.GtkBackend/MenuItemBackend.cs [hunk 11]
```diff
@@ -168,7 +168,25 @@ void ImageMenuItemStateChanged (object o, Gtk.StateChangedArgs args)
...
+			formattedText = text;
+			label.Realized += HandleStyleUpdate;
```
Comment (sevoku) [inline Xwt.Gtk/Xwt.GtkBackend/MenuItemBackend.cs]: same as in button, multiple subscription on every Markup change.

File: Xwt.Gtk/Xwt.GtkBackend/MenuItemBackend.cs [hunk 12]
```diff
@@ -249,6 +267,14 @@ void HandleItemActivated (object sender, EventArgs e)
 				context.InvokeUserCode (eventSink.OnClicked);
 			}
 		}
+
+		public void Dispose ()
+		{
+			if (label != null) {
```
Comment (sevoku) [inline Xwt.Gtk/Xwt.GtkBackend/MenuItemBackend.cs]: label never set to null, possible NRE on double dispose.

File: Xwt.XamMac/Xwt.Mac/ButtonBackend.cs [hunk 13]
```diff
@@ -153,8 +153,12 @@ public void SetButtonType (ButtonType type)
 			}
 		}
 
+		public void SetFormattedText (FormattedText text)
+		{
+			Widget.Title = text.Text;
```
Comment (sevoku) [inline Xwt.XamMac/Xwt.Mac/ButtonBackend.cs]: Is it necessary to set the Title here? It will be overwritten by AttributedTitle in the next line.

Comment (mcumming) [reply Xwt.XamMac/Xwt.Mac/ButtonBackend.cs]: Actually, this never formats the selected item in the button properly anyway, so I can remove that line.

File: Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs [hunk 14]
```diff
@@ -197,6 +199,37 @@ public void SetContent (string label, bool useMnemonic, ImageDescription image,
 				Widget.Label = null;
 		}
 		
+		FormattedText formattedText = null;
+		private void SetFormattedContent (FormattedText label, ContentPosition position)
+		{
+			if (label != null && label.Text.Length == 0)
```
Comment (Therzok) [inline Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs]: Don't we want to clear the content when label is null or it has no text?

This seems to exit when label is not null and the text is empty. Then when `labelWidget.ApplyFormattedText (formattedText)` is called, it's going to NRE inside, because formattedText is null.

File: Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs [hunk 15]
```diff
@@ -197,6 +199,37 @@ public void SetContent (string label, bool useMnemonic, ImageDescription image,
...
+				labelWidget = new Gtk.Label (label.Text);
+				Widget.Label = null;
+				Widget.Image = labelWidget;
+			}
+			labelWidget.Realized -= HandleStyleUpdate;
+			labelWidget.StyleSet -= HandleStyleUpdate;
+
+			formattedText = label;
+			labelWidget.ApplyFormattedText (formattedText);
+			labelWidget.ShowAll ();
```
Comment (Therzok) [inline Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs]: Is this ShowAll needed?

File: Xwt.Gtk/Xwt.GtkBackend/MenuItemBackend.cs [hunk 16]
```diff
@@ -168,7 +168,28 @@ void ImageMenuItemStateChanged (object o, Gtk.StateChangedArgs args)
...
-		
+
+		FormattedText formattedText = null;
+		public void SetFormattedText (FormattedText text)
+		{
+			label.Text = text?.Text;
+			formattedText = text;
+			label.Realized -= HandleStyleUpdate;
+			label.StyleSet -= HandleStyleUpdate;
+			label.ApplyFormattedText(text);
```
Comment (Therzok) [inline Xwt.Gtk/Xwt.GtkBackend/MenuItemBackend.cs]: This will NRE inside ApplyFormattedText as `text` can be null.

File: Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs [hunk 17]
```diff
@@ -130,6 +130,8 @@ public void SetContent (string label, bool useMnemonic, ImageDescription image,
 			if (image.Backend != null)
 				imageWidget = new ImageBox (ApplicationContext, image.WithDefaultSize (Gtk.IconSize.Button));
 
+			labelWidget.Realized -= HandleStyleUpdate;
```
Comment (sevoku) [inline Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs]: This NREs on first call

File: Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs (hunk 14 above)
Comment (sevoku) [reply Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs]: yep, good point

File: Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs (hunk 15 above)
Comment (sevoku) [reply Xwt.Gtk/Xwt.GtkBackend/ButtonBackend.cs]: shouldn't be needed

File: Xwt.XamMac/Xwt.Mac/MenuItemBackend.cs [hunk 18]
```diff
@@ -149,6 +149,11 @@ public bool Checked
 			}
 		}
 
+		public void SetFormattedText (FormattedText text)
+		{
+			item.AttributedTitle = text.ToAttributedString ();
```
Comment (sevoku) [inline Xwt.XamMac/Xwt.Mac/MenuItemBackend.cs]: Once set, it'll be always used, it's impossible to set an unformatted label afterwards.

File: Xwt.Gtk/Xwt.GtkBackend/MenuBackend.cs [hunk 19]
```diff
@@ -109,6 +109,7 @@ public void RemoveItem (IMenuItemBackend menuItem)
 		{
 			Gtk.MenuItem item = ((MenuItemBackend)menuItem).MenuItem;
 			menu.Remove (item);
+			item.Dispose ();
```
Comment (Therzok) [inline Xwt.Gtk/Xwt.GtkBackend/MenuBackend.cs]: Gtk code usually calls Destroy, not Dispose.

File: Xwt/Xwt/MenuItem.cs [hunk 20]
```diff
@@ -227,6 +243,12 @@ protected virtual void OnClicked (EventArgs e)
 				base.BackendHost.OnAfterEventRemove (MenuItemEvent.Clicked, clicked);
 			}
 		}
+
+		protected override void Dispose (bool release_all)
+		{
+			Backend.Dispose ();
```
Comment (Therzok) [inline Xwt/Xwt/MenuItem.cs]: We should only Dispose when `release_all` is true, otherwise we're going to be doing the following:

```
MenuItem.Finalizer() -> MacMenuItem.Dispose() -> MacMenuItem.Dispose(true)
```

Which will end up throwing an AppKit inconsistency exception for not being on the UI Thread.
